core/shared/src/main/scala/zio/ZIOConcurrent.scala

```diff
@@ -0,0 +1,153 @@
+package zio
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.ConcurrentLinkedQueue
+import scala.annotation.tailrec
+import zio.Synchronizer._
+
+final class CountDownLatch(count: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(count, maxWaiters)
+
+  // TODO is it possible to evaluate once and reuse the Promise?
+  private[this] val awaitZero = sync.acquire(count)
+
+  def await: UIO[Unit] = awaitZero.asInstanceOf[UIO[Unit]]
```

Comment:
I think we can suppress the `Full` error here but I'm not sure how to do that

Reply:
Well we need to signal the error somehow so the caller can know whether the latch has been completed and they can proceed or the latch rejected them even trying to wait. I think changing the return type to `IO[Unit, Unit]` might be the most idiomatic way to do this. We have generally shied away form putting bounds on the number of waiters just because it does push a degree of complexity onto users that can be hard to reason about but I hear your point about unbounded data structures being potential leaks.